src/parser/monk/mistweaver/modules/talents/InvokeChiJi.tsx

```diff
@@ -1,34 +1,146 @@
 import React from 'react';
 import SPELLS from 'common/SPELLS';
 import {Trans} from '@lingui/macro';
-import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
-import Events, { HealEvent } from 'parser/core/Events';
+import Analyzer, {Options, SELECTED_PLAYER, SELECTED_PLAYER_PET } from 'parser/core/Analyzer';
 import { formatNumber } from 'common/format';
+import Events, { CastEvent, HealEvent, EndChannelEvent, DeathEvent, GlobalCooldownEvent, DamageEvent} from 'parser/core/Events';
 import Statistic from 'interface/statistics/Statistic';
 import STATISTIC_ORDER from 'interface/others/STATISTIC_ORDER';
 import STATISTIC_CATEGORY from 'interface/others/STATISTIC_CATEGORY';
-import BoringSpellValueText from 'interface/statistics/components/BoringSpellValueText';
 import ItemHealingDone from 'interface/ItemHealingDone';
+import SpellLink from 'common/SpellLink';
 
+/** 
+ * Blackout Kick, Totm BoKs, Rising Sun Kick and Spinning Crane Kick generate stacks of Invoke Chi-Ji, the Red Crane, which reduce the cast time and mana 
+ * cost of Enveloping Mist by 33% per stack, up to 3 stacks.
+ * These abilities also heal 2 nearby allies for a Gust of Mist heal.
+ * Casting Enveloping Mist while Chiji is active applies Enveloping Breath on up to 6 nearby allies within 10 yards.
+*/
 class InvokeChiJi extends Analyzer {
+  MAX_STACKS: number = 3;
+  chijiActive: boolean = false;
+  //healing breakdown vars
   gustHealing: number = 0;
   envelopHealing: number = 0;
+  //stack breakdown vars
+  chijiStackCount: number = 0;
+  castsBelowMaxStacks: number = 0;
+  wastedStacks: number = 0;
+  freeCasts: number = 0;
+  //missed GCDs vars
+  missedGlobals: number = 0;
+  chijiStart: number = 0;
+  chijiGlobals: number = 0;
+  chijiUses: number = 0;
+  lastGlobal: number = 0;
+  efGcd: number = 0;
+  sckHits: string[] = [];
 
   constructor(args: Options) {
     super(args);
     this.active = this.selectedCombatant.hasTalent(SPELLS.INVOKE_CHIJI_THE_RED_CRANE_TALENT.id);
     if (!this.active) {return;}
     this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.GUST_OF_MISTS_CHIJI), this.handleGust);
     this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(SPELLS.ENVELOPING_BREATH), this.handleEnvelopingBreath);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.ENVELOPING_MIST), this.handleEnvelopCast);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SPELLS.INVOKE_CHIJI_THE_RED_CRANE_TALENT), this.handleChijiStart);
+    this.addEventListener(Events.death.to(SELECTED_PLAYER), this.handleChijiDeath);
+    this.addEventListener(Events.death.to(SELECTED_PLAYER_PET), this.handleChijiDeath);
+    this.addEventListener(Events.damage.by(SELECTED_PLAYER).spell([SPELLS.BLACKOUT_KICK, SPELLS.RISING_SUN_KICK_SECOND, SPELLS.BLACKOUT_KICK_TOTM, SPELLS.SPINNING_CRANE_KICK_DAMAGE]), this.handleStackGenerator)
```

Comment:
Okay. I'm going to double back here maybe... But while i think this will work it seems over complicated... i know log events can be weird but I think a simple cast even toggles a switch for x duration and if sck hits during that duration you count it then flip the switch off... if that makes sense. I think it would handle this a bit nice

IE
cast event -> sck{
checkForSckDamage = oldTimestamp + gcd time; (this works bc gcd = sck channel)

}
damage event -> sck {
if(checkForSckDamage>timestamp){
 send to stack counting method
checkForSckDamage = -1
}
}

Reply:
Other stuff is no problem...but what you have suggested is pretty much exactly what I had before and you told me to switch it to what it is now -_- so I'm kind of inclined to want to leave it at this point

Reply:
iirc you only had it checking if sck was cast not did damage. This is a combo of the two without making anything overly complicated :D 
You know I wouldn't do you that dirty vohrr <3

Reply:
heh it was never committed, I changed it based on this convo: 
![image](https://user-images.githubusercontent.com/26779541/96149526-69c1d400-0ed7-11eb-945c-b480a9d13e2e.png)

I'll update it, bully

Reply:
Hey. I only bully so you can learns and i can have a good product